### PR TITLE
Add IPv6 auto(enable|disable) to set NCCL use propper AF_*

### LINKED ISSUE #14: Unable run train if in system was disabled IPV6 thru kernel params.
My introduction to "multigpu" testing started from offered [Kaggle's Qwen3-Opensloth-2GPUS](https://www.kaggle.com/code/anhvth226/qwen3-opensloth-2gpus)

Form a "door" i've got this errors

```
[W729 12:37:15.952489580 socket.cpp:752] [c10d] The client socket cannot be initialized to connect to [localhost]:29501 (errno: 97 - Address family not supported by protocol).
[W729 12:37:15.041435752 socket.cpp:752] [c10d] The client socket cannot be initialized to connect to [localhost]:29501 (errno: 97 - Address family not supported by protocol).
[W729 12:37:15.044199686 CUDAAllocatorConfig.h:28] Warning: expandable_segments not supported on this platform (function operator())
[W729 12:37:15.046431838 CUDAAllocatorConfig.h:28] Warning: expandable_segments not supported on this platform (function operator())

```

And of course tried solve issue by adding some debugging in script beggining

```
test:test~$ cat train.py
import torch
import os
import socket

os.environ['NCCL_SOCKET_FAMILY'] = 'AF_INET'     # Strict use IPV4 NCCL
os.environ['MASTER_ADDR'] = '192.168.0.100'        # Address
os.environ['NCCL_SOCKET_IFNAME'] = 'eth0'          # Interface
os.environ['MASTER_PORT'] = '29500'                      # Port


print(torch.cuda.is_available())
print(torch.version.cuda)
print(os.environ.get('MASTER_ADDR'))
print(os.environ.get('MASTER_PORT'))
print(os.environ.get('NCCL_SOCKET_IFNAME'))

if torch.cuda.is_available():
    backend = 'nccl'
else:
    backend = 'gloo'

torch.distributed.init_process_group(backend=backend 

```

```
(env) test@test:~/notebook$ python train.py
True
12.4
[W729 14:26:19.507906024 socket.cpp:759] [c10d] The client socket cannot be initialized to connect to [::ffff:192.168.0.100]:29500 (errno: 97 - Address family not supported by protocol).
```

```
test:test~$ cat /etc/default/grub
GRUB_CMDLINE_LINUX_DEFAULT="ipv6.disable=1 ..........................."
```

If enable IPV6 in kernel - all goes as should. I didn't find popper OS.ENV param and this issue didn't direct related to opensloth, but now it uses NVCCL via sockets and not via MMGS.

May be your team implement "auto switch" between Memory-mapped gradient synchronization (MMGS) и NCCL-based gradient synchronization.

### PR DESCRIPTION
**Pull request by ai-bond**
It will fixes #14 and force NCCL to use IPv4 sockets only if no IPv6 support for single machine

### CODE REVIEW

**Comment by ai-bond** [comment]
Also will fixes #17 Proper collection of gradients that need synchronization and at the end of training properly clean up distributed process group.

Also fix print environment setup to get expected vars.

**Comment by ai-bond** [comment]
With same dataset before
```
┃                           ┃          ┃              ┃ Total        ┃                 ┃
┃ Step                      ┃ Count    ┃ Avg Duration ┃ Duration     ┃ Min/Max         ┃
│ model_and_training_setup  │ 1        │ 21.56s       │ 21.56s       │ 21.56s/21.56s   │
│ actual_training           │ 1        │ 4.6m         │ 4.6m         │ 4.6m/4.6m       │
│ TOTAL TRAINING            │ -        │ -            │ 4.9m         │ -               │
```
And after
```
python train.py
Global batch size: 32
Gradient accumulation steps: 16
Global batch size: 32
[MP] Running on 2 GPUs
15:55:39 | INFO     | GPU0 | opensloth_sft_trainer.py:41 | Training on GPU 0 with output_dir data/qwen3-0.6b-FineTune
15:55:39 | INFO     | GPU0 | opensloth_sft_trainer.py:44 | 🚀 Starting total training timer
🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
🦥 Unsloth Zoo will now patch everything to make training faster!
Using compiler location: .cache/unsloth_compiled_cache_0
Using compiler location: .cache/unsloth_compiled_cache_1
==((====))==  Unsloth 2025.5.7: Fast Qwen3 patching. Transformers: 4.51.3.
   \\   /|    Tesla V100-SXM2-16GB. Num GPUs = 1. Max memory: 15.766 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.7.1+cu126. CUDA: 7.0. CUDA Toolkit: 12.6. Triton: 3.3.1
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.31.post1. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
==((====))==  Unsloth 2025.5.7: Fast Qwen3 patching. Transformers: 4.51.3.
   \\   /|    Tesla V100-SXM2-16GB. Num GPUs = 1. Max memory: 15.766 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.7.1+cu126. CUDA: 7.0. CUDA Toolkit: 12.6. Triton: 3.3.1
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.31.post1. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
[W730 15:55:58.374904259 socket.cpp:755] [c10d] The client socket cannot be initialized to connect to [localhost]:29501 (errno: 97 - Address family not supported by protocol).
15:55:58 | INFO     | GPU0 | logging_config.py:161 | ⏱️  model_loading: 11.00s
[W730 15:55:58.388091492 socket.cpp:755] [c10d] The client socket cannot be initialized to connect to [localhost]:29501 (errno: 97 - Address family not supported by protocol).
[RANK=1] NCCL Environment setup: {'NCCL_SOCKET_IFNAME': 'lo', 'MASTER_ADDR': '127.0.0.1', 'MASTER_PORT': '29501'}
[RANK=0] NCCL Environment setup: {'NCCL_SOCKET_IFNAME': 'lo', 'MASTER_ADDR': '127.0.0.1', 'MASTER_PORT': '29501'}
15:55:59 | INFO     | GPU0 | init_modules.py:70 | Model loaded on device cuda:0, tokenizer: Qwen2TokenizerFast
Unsloth: Making `model.base_model.model.model` require gradients
Unsloth: Making `model.base_model.model.model` require gradients
15:56:02 | INFO     | GPU0 | logging_config.py:161 | ⏱️  lora_setup: 3.47s
15:56:02 | INFO     | GPU0 | logging_config.py:161 | ⏱️  model_init: 14.80s
15:56:02 | INFO     | GPU0 | init_modules.py:195 | Loading dataset from data/cache_qwen3_dataset/
15:56:02 | WARNING  | GPU0 | init_modules.py:172 | Dataset does not have 'labels' feature. This may affect training. Please check your dataset preparation.
[LOCAL_RANK=1] Patching log. Dir: data/qwen3-0.6b-FineTune, GPUs: 2
[LOCAL_RANK=1] Log patch initialization complete.
🔧 Patching Trainer to use RandomSamplerSeededByEpoch
[RANK=1] Patching trainer.save_model, trainer._save, and trainer._maybe_log_save_evaluate to no-op on non-master rank.
15:56:02 | INFO     | GPU0 | init_modules.py:219 | Trainer setup completed successfully
[LOCAL_RANK=0] Patching log. Dir: data/qwen3-0.6b-FineTune, GPUs: 2
[LOCAL_RANK=0] Log patch initialization complete.
🔧 Patching Trainer to use RandomSamplerSeededByEpoch
15:56:02 | INFO     | GPU0 | init_modules.py:151 | Add callback ShuffleData to Trainer UnslothSFTTrainer
15:56:02 | INFO     | GPU0 | logging_config.py:161 | ⏱️  total_setup: 23.04s
15:56:02 | INFO     | GPU0 | logging_config.py:161 | ⏱️  model_and_training_setup: 23.06s
15:56:02 | INFO     | GPU0 | opensloth_sft_trainer.py:68 | Using gradient sync callback for GPU 0
  0%|                                                                                                                                                                          | 0/31 [00:00<?, ?it/s]15:56:03 | INFO     | GPU0 | patch_sampler.py:21 | 🔄 Starting epoch 1
15:56:03 | INFO     | GPU0 | patch_sampler.py:52 | 🎲 Sampler epoch 0: emitting 1000 indices
First ids dataset samples: [776, 507, 895, 922, 33, 483, 85, 750, 354, 523]
...Last ids: [104, 754, 142, 228, 250, 281, 759, 25, 114, 654]
`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.

=== EXAMPLE #1 ===
<|im_start|>user
What is the limit of \( f(x) = 2xe^{2x} \) as \( x \) approaches positive infinity?<|im_end|>
<|im_start|>assistant
<think>
Okay, let's see. I need to find the limit of 2x times e raised to 2x as x approaches positive infinity. Hmm, let me think. So the function is f(x) = 2x e^{2x}. As x becomes really large, both 2x and e^{2x} are going to increase, right? Because e^{2x} grows exponentially, and 2x is linear. But multiplying them together, it should go to infinity. Wait, but maybe I should check if there's a way they interact. Sometimes, even if one part is growing, the other might be decreasing, but in this case, both are increasing. So intuitively, the product should go to infinity. But maybe I should be more precise.

Wait, maybe... (truncated)

More training debug examples written to .log/dataloader_examples.html
15:56:04 | INFO     | GPU0 | patch_sampler.py:28 | 📋 Dataloader examples logged to .log/dataloader_examples.html
15:56:04 | INFO     | GPU0 | patch_sampler.py:52 | 🎲 Sampler epoch 0: emitting 1000 indices
First ids dataset samples: [776, 507, 895, 922, 33, 483, 85, 750, 354, 523]
...Last ids: [104, 754, 142, 228, 250, 281, 759, 25, 114, 654]
`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.
[Step 0] Effective Tokens: 0.0M/0.0M
[Step 0] Effective Tokens: 0.0M/0.0M
{'loss': 1.0038000345230103, 'grad_norm': 0.4768220782279968, 'learning_rate': 0.0, 'epoch': 0.03}                                                                                                    
  3%|█████▏                                                                                                                                                            | 1/31 [00:17<08:51, 17.72s/it]Unsloth: Will smartly offload gradients to save VRAM!
Unsloth: Will smartly offload gradients to save VRAM!
{'loss': 1.0013999938964844, 'grad_norm': 0.4757290184497833, 'learning_rate': 2.0000000000000003e-06, 'epoch': 0.06}                                                                                 
{'loss': 0.9787999391555786, 'grad_norm': 0.5106378197669983, 'learning_rate': 4.000000000000001e-06, 'epoch': 0.09}                                                                                  
{'loss': 1.0364999771118164, 'grad_norm': 0.48162829875946045, 'learning_rate': 6e-06, 'epoch': 0.12}                                                                                                 
{'loss': 0.9620000123977661, 'grad_norm': 0.43476560711860657, 'learning_rate': 8.000000000000001e-06, 'epoch': 0.16}                                                                                 
{'loss': 1.0092999935150146, 'grad_norm': 0.43876439332962036, 'learning_rate': 1e-05, 'epoch': 0.19}                                                                                                 
{'loss': 1.0022000074386597, 'grad_norm': 0.4455691874027252, 'learning_rate': 9.615384615384616e-06, 'epoch': 0.22}                                                                                  
{'loss': 0.9754999876022339, 'grad_norm': 0.417138934135437, 'learning_rate': 9.230769230769232e-06, 'epoch': 0.25}                                                                                   
{'loss': 0.9915000200271606, 'grad_norm': 0.4095914959907532, 'learning_rate': 8.846153846153847e-06, 'epoch': 0.28}                                                                                  
{'loss': 0.9947999715805054, 'grad_norm': 0.38205188512802124, 'learning_rate': 8.461538461538462e-06, 'epoch': 0.31}                                                                                 
 32%|███████████████████████████████████████████████████▉                                                                                                             | 10/31 [01:20<02:29,  7.14s/it][Step 10] Effective Tokens: 0.2M/0.2M
[Step 10] Effective Tokens: 0.2M/0.2M
{'loss': 0.9972000122070312, 'grad_norm': 0.4048265516757965, 'learning_rate': 8.076923076923077e-06, 'epoch': 0.34}                                                                                  
{'loss': 0.9584000110626221, 'grad_norm': 0.38980498909950256, 'learning_rate': 7.692307692307694e-06, 'epoch': 0.38}                                                                                 
{'loss': 1.042799949645996, 'grad_norm': 0.3905584216117859, 'learning_rate': 7.307692307692308e-06, 'epoch': 0.41}                                                                                   
{'loss': 1.0011999607086182, 'grad_norm': 0.3940539062023163, 'learning_rate': 6.923076923076923e-06, 'epoch': 0.44}                                                                                  
{'loss': 0.990399956703186, 'grad_norm': 0.3804696798324585, 'learning_rate': 6.538461538461539e-06, 'epoch': 0.47}                                                                                   
{'loss': 1.0041999816894531, 'grad_norm': 0.3538406491279602, 'learning_rate': 6.153846153846155e-06, 'epoch': 0.5}                                                                                   
{'loss': 0.8758000135421753, 'grad_norm': 0.3455052077770233, 'learning_rate': 5.769230769230769e-06, 'epoch': 0.53}                                                                                  
{'loss': 1.0153000354766846, 'grad_norm': 0.3588210344314575, 'learning_rate': 5.384615384615385e-06, 'epoch': 0.56}                                                                                  
{'loss': 0.9409000277519226, 'grad_norm': 0.35661500692367554, 'learning_rate': 5e-06, 'epoch': 0.59}                                                                                                 
{'loss': 0.9629000425338745, 'grad_norm': 0.3431248068809509, 'learning_rate': 4.615384615384616e-06, 'epoch': 0.62}                                                                                  
 65%|███████████████████████████████████████████████████████████████████████████████████████████████████████▊                                                         | 20/31 [02:31<01:17,  7.00s/it][Step 20] Effective Tokens: 0.3M/0.3M
[Step 20] Effective Tokens: 0.3M/0.3M
{'loss': 0.9706000089645386, 'grad_norm': 0.35548099875450134, 'learning_rate': 4.230769230769231e-06, 'epoch': 0.66}                                                                                 
{'loss': 0.9485000371932983, 'grad_norm': 0.34614115953445435, 'learning_rate': 3.846153846153847e-06, 'epoch': 0.69}                                                                                 
{'loss': 0.9549000263214111, 'grad_norm': 0.3512459397315979, 'learning_rate': 3.4615384615384617e-06, 'epoch': 0.72}                                                                                 
{'loss': 0.9465999603271484, 'grad_norm': 0.35706278681755066, 'learning_rate': 3.0769230769230774e-06, 'epoch': 0.75}                                                                                
{'loss': 0.972100019454956, 'grad_norm': 0.35045570135116577, 'learning_rate': 2.6923076923076923e-06, 'epoch': 0.78}                                                                                 
{'loss': 0.9678000211715698, 'grad_norm': 0.3437287211418152, 'learning_rate': 2.307692307692308e-06, 'epoch': 0.81}                                                                                  
{'loss': 0.9390000104904175, 'grad_norm': 0.3487148582935333, 'learning_rate': 1.9230769230769234e-06, 'epoch': 0.84}                                                                                 
{'loss': 1.000599980354309, 'grad_norm': 0.34078580141067505, 'learning_rate': 1.5384615384615387e-06, 'epoch': 0.88}                                                                                 
{'loss': 1.0390000343322754, 'grad_norm': 0.3489170968532562, 'learning_rate': 1.153846153846154e-06, 'epoch': 0.91}                                                                                  
{'loss': 0.8924000263214111, 'grad_norm': 0.3260272145271301, 'learning_rate': 7.692307692307694e-07, 'epoch': 0.94}                                                                                  
 97%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊     | 30/31 [03:43<00:07,  7.19s/it][Step 30] Effective Tokens: 0.5M/0.5M
[Step 30] Effective Tokens: 0.5M/0.5M
{'loss': 0.9510999917984009, 'grad_norm': 0.3215421736240387, 'learning_rate': 3.846153846153847e-07, 'epoch': 0.97}                                                                                  
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 31/31 [03:50<00:00,  7.45s/it]
{'train_runtime': 232.1152, 'train_samples_per_second': 4.308, 'train_steps_per_second': 0.134, 'train_loss': 0.49629371396956906, 'epoch': 0.97}                                                     
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 31/31 [03:52<00:00,  7.49s/it]
[GPU 1] Distributed process group destroyed.
[GPU 0] Distributed process group destroyed.
15:59:55 | INFO     | GPU0 | logging_config.py:161 | ⏱️  actual_training: 3.9m
                            ⏱️  Training Step Timing Summary                             
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃                           ┃          ┃              ┃ Total        ┃                 ┃
┃ Step                      ┃ Count    ┃ Avg Duration ┃ Duration     ┃ Min/Max         ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ model_and_training_setup  │ 1        │ 23.06s       │ 23.06s       │ 23.06s/23.06s   │
│ actual_training           │ 1        │ 3.9m         │ 3.9m         │ 3.9m/3.9m       │
│ TOTAL TRAINING            │ -        │ -            │ 4.3m         │ -               │
└───────────────────────────┴──────────┴──────────────┴──────────────┴─────────────────┘
All processes finished
```
